### PR TITLE
[DAG] Add legalization handling for AVGCEIL/AVGFLOOR nodes

### PR DESCRIPTION
**Pull request by RKSimon**
Always match AVG patterns pre-legalization, and use TargetLowering::expandAVG to expand again during legalization.

I've removed the X86 custom AVGCEILU pattern detection and replaced with combines to try and convert other AVG nodes to AVGCEILU.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/TargetLowering.h [hunk 1]
```diff
@@ -5332,6 +5332,11 @@ class TargetLowering : public TargetLoweringBase {
   /// \returns The expansion result or SDValue() if it fails.
   SDValue expandABD(SDNode *N, SelectionDAG &DAG) const;
 
+  /// Expand vector/scalar AVGCEILS/AVGCEILU/AVGFLOOR/AVGFLOORU nodes.
```
Comment (jayfoad) [inline llvm/include/llvm/CodeGen/TargetLowering.h:5335]:
```suggestion
  /// Expand vector/scalar AVGCEILS/AVGCEILU/AVGFLOORS/AVGFLOORU nodes.
```

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -4752,6 +4752,19 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
...
+    // Treat the AVG nodes like ADD/SUB nodes as that's the only stage that
+    // can lose sign bits:
+    // avgceils(lhs, rhs) -> sub(or(lhs,rhs),ashr(xor(lhs,rhs),1))
+    // avgfloors(lhs, rhs) -> add(and(lhs,rhs),ashr(xor(lhs,rhs),1))
+    Tmp = ComputeNumSignBits(Op.getOperand(0), DemandedElts, Depth + 1);
+    if (Tmp == 1)
+      break; // Early out.
+    Tmp2 = ComputeNumSignBits(Op.getOperand(1), DemandedElts, Depth + 1);
+    if (Tmp2 == 1)
+      break; // Early out.
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4766]: Remove this? The expression on the next line is not expensive.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 3]
```diff
@@ -4752,6 +4752,19 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
...
+    return std::min(Tmp, Tmp2) - 1;
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4767]: Intuitively I would hope that we don't need the `- 1` here but I'm not sure. Do we have any exhaustive tests for NumSignBits calculations?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 2 above)
Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4766]: Its to prevent the sign bits result equaling zero

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 3 above)
Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4767]: No, I'm starting to wonder if we should have some form of SignBits equivalent to KnownBits so we can share more valuetracking code - what do you think?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4767]: Not exhaustive, but random fuzzing suggests we can use `std::min(Tmp,Tmp2)` - https://godbolt.org/z/Trx1986rj

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll [hunk 4]
```diff
@@ -39,8 +39,10 @@ define <8 x i8> @vaaddu_vv_v8i8_floor_sexti16(<8 x i8> %x, <8 x i8> %y) {
 ; CHECK-LABEL: vaaddu_vv_v8i8_floor_sexti16:
 ; CHECK:       # %bb.0:
 ; CHECK-NEXT:    vsetivli zero, 8, e8, mf2, ta, ma
-; CHECK-NEXT:    vwadd.vv v10, v8, v9
-; CHECK-NEXT:    vnsrl.wi v8, v10, 1
+; CHECK-NEXT:    vand.vv v10, v8, v9
+; CHECK-NEXT:    vxor.vv v8, v8, v9
+; CHECK-NEXT:    vsra.vi v8, v8, 1
+; CHECK-NEXT:    vadd.vv v8, v10, v8
```
Comment (RKSimon) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:45]: @topperc Would it be acceptable to just add avgfloors/avgflooru custom lowering to riscv? I'm not sure how better we can make use the vwadd widening add instruction.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 3 above)
Comment (jayfoad) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: > No, I'm starting to wonder if we should have some form of SignBits equivalent to KnownBits so we can share more valuetracking code - what do you think?

We could create something similar to KnownBits that calculates numsignbits for an operation based on the numsignbits of its inputs, but I'm not sure how useful that would be. My hunch is that most of the useful stuff that happens in numsignbits calculations relies on more information than just the numsignbits of the inputs.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 5]
```diff
@@ -5234,6 +5234,20 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
   if (N0 == N1 && Level >= AfterLegalizeTypes)
     return N0;
 
+  // Fold avgflooru(x,c) -> avgceilu(x,c-1)
+  // iff c > 0 and avgflooru isn't legal/custom but avgceilu is.
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5238]: Is this a quirk of some specific target? Otherwise, why not do this for the signed cases too, and do the opposite fold from ceil -> floor?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5238]: Its purely to get around x86 regressions (which only has avgceilu) - we can add other equivalences if there are other targets don't have all the avg variants.

Comment (jayfoad) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5238]: > Its purely to get around x86 regressions (which only has avgceilu)

Ack. Is it worth legalizing the signed versions to the unsigned ones by flipping the sign bit of both inputs and the result?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5238]: Yes, that's another one that is probably worth it (not sure if we have an existing test case for that, although the avg tests are more extensive than you'd expect.....). These are pretty generic folds, but I can move this into X86 if people prefer?

Comment (goldsteinn) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5238]: Unless there is certainty that no other target can benefit, imo its written generically enough to keep here.

I.e the general pattern: `if(NotLegal(A) && IsLegal(B) && MayTransform(A, B)) { return B; }` is sensible as a generic combine, even if the motivation was one particular target.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 6]
```diff
@@ -4560,8 +4560,15 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
...
-            getValidMinimumShiftAmountConstant(Op, DemandedElts))
+            getValidMinimumShiftAmountConstant(Op, DemandedElts)) {
       Tmp = std::min<uint64_t>(Tmp + ShAmt->getZExtValue(), VTBits);
+    } else {
+      KnownBits KnownAmt =
+          computeKnownBits(Op.getOperand(1), DemandedElts, Depth + 1);
+      if (KnownAmt.isConstant() && KnownAmt.getConstant().ult(VTBits))
+        Tmp = std::min<uint64_t>(Tmp + KnownAmt.getConstant().getZExtValue(),
+                                 VTBits);
+    }
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4571]: This seems like an unrelated change?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4571]: If only... its to fix a thumb2 regression as it lowers v2i64 constant as bitcast(v4i32 constant)

Once this draft has addressed all the regressions I'll turn my attention to pulling out some of these changes.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll (hunk 4 above)
Comment (RKSimon) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:45]: @topperc @preames These are the last outstanding regressions in the draft patch before I start cleanup - do you have any suggestions on the preferred way that riscv should handle AVGFLOORS? AVGCEILS appears to be fine.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:45]: We can add custom lowering. Ideally we'd use vaadd, but that requires a write to a rounding mode register that some microarchitectures haven't optimized.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:45]: Oh I guess we do use vaaddu for some cases already. I guess we just need to do the rest.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:45]: Does AVGFLOORS of -1 and 0 return -1? I think that's what stopped this from getting implemented before. There was maybe confusion around the meaning of "average". https://github.com/riscv/riscv-v-spec/issues/935

Comment (RKSimon) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll:45]: Yes, it should return -1:
```
avgfloors(lhs, rhs) -> add(and(lhs,rhs),ashr(xor(lhs,rhs),1))
avgfloors(lhs, rhs) -> trunc(ashr(add(sext(lhs),sext(rhs)),1))
avgfloors(lhs, rhs) -> trunc(lshr(add(sext(lhs),sext(rhs)),1))
```

Comment (RKSimon) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-vaaddu.ll]: Thanks @topperc

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp (hunk 6 above)
Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4599]: This is proving tricky to pull out - but I've confirmed that it doesn't cause any notable [compile time diff](https://llvm-compile-time-tracker.com/compare.php?from=7630379156ec08c9d7b1ea3c03c09e7dc89ef4ee&to=7a82854b3d28d02d37183ea91f09b1f7343c4bc4&stat=instructions:u) - as we fallback to ComputeKnownBits call which will call computeKnownBits on the shift amount anyway.

Comment (jayfoad) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4599]: Isn't there any helper simpler than computeKnownBits that can look through bitcasts to find a constant?

If you _are_ going to use computeKnownBits, why not use `KnownAmt.getMinValue()` instead of `KnownAmt.getConstant()`?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4599]: Updated to getMaxValue() (for upper bound) + getMinValue() (for min sign extension) - the shift amount isn't just a `bitcast(v4i32 constant)` hidden constant, so we do need the abilities of computeKnownBits.

We could update getValidMinimumShiftAmountConstant (et. al) to return `std::optional<APInt>` to allow it to fallback to computeKnownBits, although that would mean the function would return a value that might not actual exist in the shift amount, I don't think we've used that property but it would still be a change.

Comment (jayfoad) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4599]: Actually I'm not even sure what the `ult` check is for, except perhaps to guard against `Tmp + getMinValue()` overflowing?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:4599]: Yes, its mainly just a sanity/overflow check (somebody always comes along with a i1024 fuzz test or something eventually that makes getZExtValue() assert or cause weird getLimitedValue() behaviour).

Using getMaxValue() was mainly to try and keep closer to the behaviour of getValidMinimumShiftAmountConstant which doesn't accept out of bounds shift amounts.

**Comment by RKSimon** [comment]
ping - #93182 is now finished, so this PR is ready to go.

**Comment by RKSimon** [comment]
ping? any objections to me getting this committed now please?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 7]
```diff
@@ -5241,6 +5241,21 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
     return DAG.getNode(ISD::SRL, DL, VT, X,
                        DAG.getShiftAmountConstant(1, VT, DL));
 
+  // Fold avgflooru(x,y) -> avgceilu(x,y-1) iff y != 0
+  // Fold avgflooru(x,y) -> avgceilu(x-1,y) iff x != 0
+  // Check if avgflooru isn't legal/custom but avgceilu is.
+  if (Opcode == ISD::AVGFLOORU && !hasOperation(ISD::AVGFLOORU, VT) &&
+      (!LegalOperations || hasOperation(ISD::AVGCEILU, VT))) {
+    if (DAG.isKnownNeverZero(N0))
```
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5249]: Maybe handle N1 before N0? I assume if either one is a constant it will have been canonicalized to the RHS, and you should prefer to subtract 1 from a constant.

**Comment by RKSimon** [comment]
@jayfoad any more comments?

**Review by jayfoad — COMMENTED** [review]
No objection from me. The logic looks good. But I don't feel I know enough about any of the affected targets to approve it.

**Comment by RKSimon** [comment]
@davemgreen  @goldsteinn any objections?

**Comment by RKSimon** [comment]
ping?

**Review by jayfoad — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
Hi @RKSimon, I think this patch causes some regressions on riscv: https://github.com/dtcxzyw/llvm-codegen-benchmark/commit/97ad8e7b80da7aba30042b112095724802189681

Reproducer:
```
; llc -mtriple=riscv64 test.ll -o -
define signext i64 @func000000000000002b(i32 signext %0) #0 {
entry:
  %1 = zext nneg i32 %0 to i64
  %2 = add nsw i64 %1, -1
  %3 = lshr i64 %2, 1
  %4 = add nuw nsw i64 %3, 1
  %5 = and i64 %4, 9223372036854775806
  ret i64 %5
}
```
Before (74f200baedfed496880ca86ce9409788b0d0eaca):
```
func000000000000002b:
        addi    a0, a0, -1
        srli    a0, a0, 1
        addi    a0, a0, 1
        andi    a0, a0, -2
        ret
```
After (47afa10bbaa89351afa9bcc53dd959e6181ebf3d):
```
func000000000000002b:
        addi    a0, a0, -1
        srli    a0, a0, 1
        addi    a0, a0, 1
        li      a1, -3
        srli    a1, a1, 1
        and     a0, a0, a1
        ret
```

**Comment by dtcxzyw** [comment]
> Hi @RKSimon, I think this patch causes some regressions on riscv: [dtcxzyw/llvm-codegen-benchmark@97ad8e7](https://github.com/dtcxzyw/llvm-codegen-benchmark/commit/97ad8e7b80da7aba30042b112095724802189681)
> 
> Reproducer:
> 
> ```
> ; llc -mtriple=riscv64 test.ll -o -
> define signext i64 @func000000000000002b(i32 signext %0) #0 {
> entry:
>   %1 = zext nneg i32 %0 to i64
>   %2 = add nsw i64 %1, -1
>   %3 = lshr i64 %2, 1
>   %4 = add nuw nsw i64 %3, 1
>   %5 = and i64 %4, 9223372036854775806
>   ret i64 %5
> }
> ```
> 
> Before ([74f200b](https://github.com/llvm/llvm-project/commit/74f200baedfed496880ca86ce9409788b0d0eaca)):
> 
> ```
> func000000000000002b:
>         addi    a0, a0, -1
>         srli    a0, a0, 1
>         addi    a0, a0, 1
>         andi    a0, a0, -2
>         ret
> ```
> 
> After ([47afa10](https://github.com/llvm/llvm-project/commit/47afa10bbaa89351afa9bcc53dd959e6181ebf3d)):
> 
> ```
> func000000000000002b:
>         addi    a0, a0, -1
>         srli    a0, a0, 1
>         addi    a0, a0, 1
>         li      a1, -3
>         srli    a1, a1, 1
>         and     a0, a0, a1
>         ret
> ```

```
SelectionDAG has 17 nodes:
  t0: ch,glue = EntryToken
                  t2: i64,ch = CopyFromReg t0, Register:i64 %0
                t4: i64 = AssertSext t2, ValueType:ch:i32
              t5: i32 = truncate t4
            t6: i64 = sign_extend t5
          t8: i64 = add nsw t6, Constant:i64<-1>
        t10: i64 = srl t8, Constant:i64<1>
      t11: i64 = add nuw nsw t10, Constant:i64<1>
    t13: i64 = and t11, Constant:i64<9223372036854775806>
  t15: ch,glue = CopyToReg t0, Register:i64 $x10, t13
  t16: ch = RISCVISD::RET_GLUE t15, Register:i64 $x10, t15:1



Combining: t16: ch = RISCVISD::RET_GLUE t15, Register:i64 $x10, t15:1

Combining: t15: ch,glue = CopyToReg t0, Register:i64 $x10, t13

Combining: t14: i64 = Register $x10

Combining: t13: i64 = and t11, Constant:i64<9223372036854775806>
Creating constant: t17: i32 = Constant<-1>
Creating new node: t18: i32 = avgfloors t5, Constant:i32<-1>
Creating new node: t19: i64 = sign_extend t18
```

**Comment by RKSimon** [comment]
cheers - looking at this now

**Comment by vitalybuka** [comment]
Probably by this patch as this is the only one in DAG in the blame list
https://lab.llvm.org/buildbot/#/builders/237/builds/7908

Can you please fix or revert?

FYI @fmayer

**Comment by dtcxzyw** [comment]
> Probably by this patch as this is the only one in DAG in the blame list https://lab.llvm.org/buildbot/#/builders/237/builds/7908
> 
> Can you please fix or revert?
> 
> FYI @fmayer

Should be fixed by https://github.com/llvm/llvm-project/commit/ca33796d54ce6d2c711032b269caf32851c5915a.
